web/src/settings_components.ts

```diff
@@ -359,6 +359,71 @@ function get_message_retention_setting_value(
     return check_valid_number_input(custom_input_val);
 }
 
+type FieldData = Record<string, Record<string, string> | string>;
+
+function read_select_field_data_from_form(
+    $profile_field_form: JQuery,
+    old_field_data: FieldData,
+): FieldData {
+    const field_data: FieldData = {};
+    let field_order = 1;
+
+    const old_option_value_map = new Map();
+    if (old_field_data !== undefined) {
+        for (const [value, choice] of Object.entries(old_field_data)) {
+            assert(typeof choice !== "string");
+            old_option_value_map.set(choice.text, value);
+        }
+    }
+    $profile_field_form.find("div.choice-row").each(function (this: HTMLElement) {
+        const text = $(this).find("input")[0].value;
+        if (text) {
+            if (old_option_value_map.get(text) !== undefined) {
+                // Resetting the data-value in the form is
+                // important if the user removed an option string
+                // and then added it back again before saving
+                // changes.
+                $(this).attr("data-value", old_option_value_map.get(text));
+            }
+            const value = $(this).attr("data-value")!;
+            field_data[value] = {text, order: field_order.toString()};
+            field_order += 1;
+        }
+    });
+
+    return field_data;
+}
+
+function read_external_account_field_data($profile_field_form: JQuery): FieldData {
+    const field_data: FieldData = {};
+    field_data.subtype = $profile_field_form
+        .find<HTMLSelectElement & {type: "select-one"}>("select[name=external_acc_field_type]")
```

Comment:
This `HTMLSelectElement & {type: "select-one"}` type may be handy; I think we have some other TS files that do something fairly messy to assert that the value is a string. Might be worth having an `HTMLSelectSingleElement` type of this form that we use more widely; probably a good #frontend conversation to start.

Reply:
Opened discussion at https://chat.zulip.org/#narrow/stream/6-frontend/topic/Type.20of.20HTMLSelectElement.20with.20single.20value.2E/near/1789464